src/Data/LLVM/BitCode/GetBits.hs

```diff
@@ -125,112 +117,233 @@ instance MonadPlus GetBits where
 -- extract is not valid... i.e. > bitLimit).  Returns the Integer value along
 -- with the bit position following the extraction.
 
-extractFromByteString :: NumBits {-^ the last bit accessible in the ByteString -}
-                      -> NumBits {-^ the bit to start extraction at -}
-                      -> NumBits {-^ the number of bits to extract -}
+-- There are two implementations: one builds an integer value by shifting bits,
+-- then shifts and masks the result to get the final value.  The other uses
+-- unlifted values and avoids the final shift by being smarter about individual
+-- compositions.  Their functionality should be identical, but it may be easier
+-- to debug the first.
```

Comment:
Perhaps it's worth defining the simpler version under a different name. It wouldn't actually be used anywhere, but that way the code would still be typechecked, making it less susceptible to bitrot.

Reply:
Good idea.